### PR TITLE
Importer cleanup

### PR DESCRIPTION
**Pull request by reuben**
This is the big importer refactor. The commits are organized in a way that the repository still works after each one is applied. The largest changes are in commit d1fabe0bbca738981c1840d3b6a278787ef4f83b, the first one, which tries to do a single pass over the corpora where possible. In practice, this means audio and transcriptions are split in the same pass. In addition, it also adds code to save the dataset definition in CSV files, and read/handle them using [pandas](http://pandas.pydata.org/).

Commit a88e4ae2f31bc45d073c6e57adc618c6ce5468bd simply refactors the DataSet class into util/data_set_helpers.py, as all the importers have identical implementations now.

Then comes commit c4ed81ce43d52ff3674ee3d49626acb888420120, which makes the main script read the .csv files directly instead of calling into the importer.

Finally, commit 6133ef3b70813c77d2f9d182fbc1944168a056fb changes the importers to be scripts that work independently from the main script. The work flow from zero to training on say, LibriSpeech, is now something like:
```
python util/importers/librivox.py path/to/destination/folder
python DeepSpeech.py --train_files path/to/destination/folder/librivox-train-clean-100.csv --dev_files path/to/destination/folder/librivox-dev-clean.csv --test_files path/to/destination/folder/librivox-test.clean.csv
```

I'm gonna add pandas to requirements.txt and update the README tomorrow. I tested this code with LDC93S1 and LibriSpeech, but couldn't test with TED-LIUM as the archive on the server seems to be corrupted.

### CODE REVIEW

**Comment by reuben** [comment]
@tilmankamp I'd like you to do a full review as I think you'll be the first one to build on top of this with audio augmentation for example.

@kdavis-mozilla I'd like you to review the UI/API changes in particular, and the changes to the TED-LIUM importer.

@andrenatal I'd like you to review the changes to the Switchboard importer in particular.

I just realized I forgot to update `bin/run-wer-automation.sh`. I'm gonna do that tomorrow with the README/requirements.txt changes.

File: util/importers/ldc93s1.py [hunk 1]
```diff
@@ -1,113 +1,29 @@
...
-
-        with open(txt_file) as open_txt_file:
-            original = ' '.join(open_txt_file.read().strip().lower().split(' ')[2:]).replace('.', '')
-
-        target = text_to_char_array(original)
+# Make sure we can import stuff from util/
+# This script needs to be run from the root of the DeepSpeech repository
+import sys
+import os
+sys.path.insert(1, os.path.join(sys.path[0], '..', '..'))
```
Comment (tilmankamp) [inline util/importers/ldc93s1.py]: Quite hackish - but if this is common practice, I don't mind.

File: util/importers/librivox.py [hunk 2]
```diff
@@ -1,124 +1,27 @@
-from __future__ import print_function
-from __future__ import absolute_import
+#!/usr/bin/env python
+from __future__ import absolute_import, division, print_function
+
+# Make sure we can import stuff from util/
+# This script needs to be run from the root of the DeepSpeech repository
+import sys
+import os
+sys.path.insert(1, os.path.join(sys.path[0], '..', '..'))
```
Comment (tilmankamp) [inline util/importers/librivox.py]: Quite hackish - but if this is common practice, I don't mind.

File: util/importers/ted.py [hunk 3]
```diff
@@ -1,123 +1,26 @@
...
-import wave
-import random
+#!/usr/bin/env python
+from __future__ import absolute_import, division, print_function
+
+# Make sure we can import stuff from util/
+# This script needs to be run from the root of the DeepSpeech repository
+import sys
+import os
+sys.path.insert(1, os.path.join(sys.path[0], '..', '..'))
```
Comment (tilmankamp) [inline util/importers/ted.py]: Quite hackish - but if this is common practice, I don't mind.

File: util/importers/LDC97S62.py [hunk 4]
```diff
@@ -1,112 +1,21 @@
...
-import fnmatch
-import numpy as np
+#!/usr/bin/env python
+from __future__ import absolute_import, division, print_function
+
+# Make sure we can import stuff from util/
+# This script needs to be run from the root of the DeepSpeech repository
+import sys
 import os
+sys.path.insert(1, os.path.join(sys.path[0], '..', '..'))
```
Comment (tilmankamp) [inline util/importers/LDC97S62.py]: Quite hackish - but if this is common practice, I don't mind.

File: util/importers/fisher.py [hunk 5]
```diff
@@ -1,153 +1,58 @@
-from __future__ import print_function
-from __future__ import absolute_import
+#!/usr/bin/env python
+from __future__ import absolute_import, division, print_function
+
+# Make sure we can import stuff from util/
+# This script needs to be run from the root of the DeepSpeech repository
+import sys
+import os
+sys.path.insert(1, os.path.join(sys.path[0], '..', '..'))
```
Comment (tilmankamp) [inline util/importers/fisher.py]: Quite hackish - but if this is common practice, I don't mind.

File: util/stm.py [hunk 6]
```diff
@@ -13,7 +15,9 @@ def __init__(self, stm_line):
         self._transcript  = ""
         for token in tokens[6:]:
           self._transcript += token + " "
-        self._transcript = self._transcript.strip()
+        self._transcript = unicodedata.normalize("NFKD", self._transcript.strip())  \
+                                      .encode("ascii", "ignore")                    \
+                                      .decode("ascii", "ignore")
```
Comment (tilmankamp) [inline util/stm.py]: What is this encode/decode step for?

File: util/text.py [hunk 7]
```diff
@@ -201,5 +201,5 @@ def validate_label(label):
     label = label.replace("?", "")
     label = label.strip()
 
-    return label
+    return label.lower()
```
Comment (tilmankamp) [inline util/text.py]: For most languages ok, how about moving this to the importer?

File: util/importers/LDC97S62.py [hunk 8]
```diff
@@ -1,39 +1,35 @@
...
 import unicodedata
-import codecs
-from glob import glob
-from itertools import cycle
+import wave
+
 from math import ceil
-from Queue import PriorityQueue
-from Queue import Queue
+from six.moves import range
```
Comment (kdavis-mozilla) [inline util/importers/LDC97S62.py]: Do we really need to import `range` here or can we simply rely on the python 2.x or python 3.x `range`?

We only use `range` here
```python
batch_threads = [Thread(target=self._populate_batch_queue, args=(session,)) for i in range(self._thread_count)]
```
and `self._thread_count` isn't large.  So I don't see any harm in simply falling back to python 2.x or python 3.x `range`.

Is there some problem with my logic?

File: util/importers/LDC97S62.py [hunk 9]
```diff
@@ -1,39 +1,35 @@
...
-        self._y = tf.placeholder(tf.int32, [None, ])
+        self._y = tf.placeholder(tf.int32, [None,])
         self._y_length = tf.placeholder(tf.int32, [])
         self.example_queue = tf.PaddingFIFOQueue(shapes=[[None, numcep + (2 * numcep * numcontext)], [], [None, ], []],
                                                   dtypes = [tf.float32, tf.int32, tf.int32, tf.int32],
                                                   capacity = 2 * self._get_device_count() * batch_size)
         self._enqueue_op = self.example_queue.enqueue([self._x, self._x_length, self._y, self._y_length])
         self._close_op = self.example_queue.close(cancel_pending_enqueues=True)
-        self._txt_files = txt_files
+        self._filelist = filelist
```
Comment (kdavis-mozilla) [inline util/importers/LDC97S62.py]: Maybe remove this member? Having both the name `_filelist` and `_files_list` is a bit confusing.

An idea would be to 
- Remove `_filelist` completely
- Pass `filelist` to `_create_files_list()`
- Use `_files_list` in `total_batches()`

File: util/importers/LDC97S62.py [hunk 10]
```diff
@@ -102,50 +87,70 @@ def next_batch(self):
...
-                     "swb1_d3-split-wav")
-    _maybe_split_wav(data_dir, "swb_ms98_transcriptions", "swb1_d4-wav",
-                     "swb1_d4-split-wav")
-
-    _maybe_split_transcriptions(data_dir, "swb_ms98_transcriptions")
-
-    _maybe_split_sets(data_dir, ["swb1_d1-split-wav", "swb1_d2-split-wav", "swb1_d3-split-wav", "swb1_d4-split-wav"],
-                      "final_sets")
+        # Note: If len(_filelist) % batch_size != 0, this re-uses initial files
+        return int(ceil(len(self._filelist) / self.batch_size))
```
Comment (kdavis-mozilla) [inline util/importers/LDC97S62.py]: The import
```python
from __future__ import absolute_import, division, print_function
```
changes the value returned here.

For example if `len(self._filelist)` were 10 and `self.batch_size` were 6, then in python 2.x one finds
```bash
>>> from math import ceil
>>> int(ceil(10 / 6))
1
>>> from __future__ import division
>>> int(ceil(10 / 6))
2
```
So this changes the `total_batches` property.

File: util/importers/LDC97S62.py [hunk 11]
```diff
@@ -175,36 +180,32 @@ def _maybe_convert_wav(data_dir, original_data, converted_data):
...
-            stop_time_end = line.find(" ", stop_time_beg)
-
-            transcript_beg = stop_time_end + 1
-            transcript_end = len(line)
-
-            if validate_label(line[transcript_beg:transcript_end].strip()) == None:
+            if transcript == None:
                 continue
 
+            transcript = unicodedata.normalize("NFKD", transcript)  \
```
Comment (kdavis-mozilla) [inline util/importers/LDC97S62.py]: Could you include the original comment on why this encode/decode dance is required?

File: util/data_set_helpers.py [hunk 12]
```diff
@@ -1,8 +1,12 @@
+from __future__ import absolute_import, division, print_function
```
Comment (kdavis-mozilla) [inline util/data_set_helpers.py]: Are you sure the `division` import does not change the behaviour of any code in this file?

File: util/data_set_helpers.py [hunk 13]
```diff
@@ -1,8 +1,12 @@
...
 
-import threading
-import numpy as np
 import tensorflow as tf
 
+from threading import Thread
+from math import ceil
+from six.moves import range
```
Comment (kdavis-mozilla) [inline util/data_set_helpers.py]: Do we really need to import `range` here or can we simply rely on the python 2.x or python 3.x `range`?

File: util/data_set_helpers.py [hunk 14]
```diff
@@ -35,6 +39,80 @@ def dev(self):
...
+        self.example_queue = tf.PaddingFIFOQueue(shapes=[[None, numcep + (2 * numcep * numcontext)], [], [None,], []],
+                                                  dtypes=[tf.float32, tf.int32, tf.int32, tf.int32],
+                                                  capacity=2 * self._get_device_count() * batch_size)
+        self._enqueue_op = self.example_queue.enqueue([self._x, self._x_length, self._y, self._y_length])
+        self._close_op = self.example_queue.close(cancel_pending_enqueues=True)
+        self._filelist = filelist
+        self.batch_size = batch_size
+        self._numcontext = numcontext
+        self._thread_count = thread_count
+        self._files_list = self._create_files_list()
```
Comment (kdavis-mozilla) [inline util/data_set_helpers.py]: Maybe name this `_list_of_files`? Having both the name `_filelist` and `_files_list` is a bit confusing.

Another idea would be to 
- Remove `_filelist` completely
- Pass `filelist` to `_create_files_list()`
- Use `_files_list` in `total_batches()`

File: util/data_set_helpers.py [hunk 15]
```diff
@@ -35,6 +39,80 @@ def dev(self):
...
+
+    def next_batch(self):
+        source, source_lengths, target, target_lengths = self.example_queue.dequeue_many(self.batch_size)
+        sparse_labels = ctc_label_dense_to_sparse(target, target_lengths, self.batch_size)
+        return source, source_lengths, sparse_labels
+
+    @property
+    def total_batches(self):
+        # Note: If len(_filelist) % batch_size != 0, this re-uses initial files
+        return int(ceil(len(self._filelist) / self.batch_size))
```
Comment (kdavis-mozilla) [inline util/data_set_helpers.py]: The import
```python
from __future__ import absolute_import, division, print_function
```
changes the value returned here.

For example if `len(self._filelist)` were 10 and `self.batch_size` were 6, then in python 2.x one finds
```bash
>>> from math import ceil
>>> int(ceil(10 / 6))
1
>>> from __future__ import division
>>> int(ceil(10 / 6))
2
```
So this changes the `total_batches` property.

File: DeepSpeech.py [hunk 16]
```diff
@@ -9,30 +9,23 @@
...
-import tempfile
 import tensorflow as tf
 import time
-import importlib
-import cgi
-import pickle
 
-from collections import OrderedDict
-from math import ceil, floor
+from six.moves import zip, range, filter, urllib, BaseHTTPServer
```
Comment (kdavis-mozilla) [inline DeepSpeech.py]: Why do we need this `range`? Couldn't we default to the Python 2.X or 3.X `range`?

File: bin/run-fisher.sh [hunk 17]
```diff
@@ -5,8 +5,22 @@ if [ ! -f DeepSpeech.py ]; then
...
+    ds_dataroot = "data"
+fi;
+
+# Warn if we can't find the train files
+if [ ! -f "${ds_dataroot}/fisher-train.csv" ]; then
+    echo "Warning: It looks like you don't have the Fisher corpus "            \
+         "downloaded and preprocessed. Make sure \$ds_dataroot points to the " \
+         "folder where the Fisher data is located, and that you ran the "      \
+         "importer script before running this script."
+fi;
```
Comment (kdavis-mozilla) [inline bin/run-fisher.sh]: I agree on adding the file name of the importer script here. It'll save us many a random issue filed by contributors.

File: bin/run-librivox.sh [hunk 18]
```diff
@@ -5,10 +5,26 @@ if [ ! -f DeepSpeech.py ]; then
...
+    ds_dataroot = "data"
+fi;
 
+# Warn if we can't find the train files
+if [ ! -f "${ds_dataroot}/librivox-train-clean-100.csv" ]; then
+    echo "Warning: It looks like you don't have the LibriSpeech corpus "       \
+         "downloaded and preprocessed. Make sure \$ds_dataroot points to the " \
+         "folder where the LibriSpeech data is located, and that you ran the " \
+         "importer script before running this script."
+fi;
```
Comment (kdavis-mozilla) [inline bin/run-librivox.sh]: I agree on adding the file name of the importer script here. It'll save us many a random issue filed by contributors.

File: bin/run-swb.sh [hunk 19]
```diff
@@ -5,8 +5,22 @@ if [ ! -f DeepSpeech.py ]; then
...
+    ds_dataroot = "data"
+fi;
+
+# Warn if we can't find the train files
+if [ ! -f "${ds_dataroot}/swb-train.csv" ]; then
+    echo "Warning: It looks like you don't have the Switchboard corpus "       \
+         "downloaded and preprocessed. Make sure \$ds_dataroot points to the " \
+         "folder where the Switchboard data is located, and that you ran the " \
+         "importer script before running this script."
+fi;
```
Comment (kdavis-mozilla) [inline bin/run-swb.sh]: I agree on adding the file name of the importer script here. It'll save us many a random issue filed by contributors.

File: bin/run-ted.sh [hunk 20]
```diff
@@ -6,8 +6,22 @@ if [ ! -f DeepSpeech.py ]; then
...
+    ds_dataroot = "data"
+fi;
+
+# Warn if we can't find the train files
+if [ ! -f "${ds_dataroot}/ted-train.csv" ]; then
+    echo "Warning: It looks like you don't have the TED-LIUM corpus "          \
+         "downloaded and preprocessed. Make sure \$ds_dataroot points to the " \
+         "folder where the TED-LIUM data is located, and that you ran the " \
+         "importer script before running this script."
+fi;
```
Comment (kdavis-mozilla) [inline bin/run-ted.sh]: I agree on adding the file name of the importer script here. It'll save us many a random issue filed by contributors.

File: util/importers/ldc93s1.py [hunk 21]
```diff
@@ -1,61 +1,29 @@
+#!/usr/bin/env python
```
Comment (kdavis-mozilla) [inline util/importers/ldc93s1.py]: Maybe move this to bin and rename it something like `import_ldc93s1.py`?

File: util/importers/ted.py [hunk 22]
```diff
@@ -1,9 +1,15 @@
+#!/usr/bin/env python
```
Comment (kdavis-mozilla) [inline util/importers/ted.py]: Maybe move this to bin and rename it something like import_ted.py?

File: util/importers/librivox.py [hunk 23]
```diff
@@ -1,68 +1,23 @@
+#!/usr/bin/env python
```
Comment (kdavis-mozilla) [inline util/importers/librivox.py]: Maybe move this to bin and rename it something like import_librivox.py?

File: util/importers/fisher.py [hunk 24]
```diff
@@ -1,91 +1,58 @@
+#!/usr/bin/env python
```
Comment (kdavis-mozilla) [inline util/importers/fisher.py]: Maybe move this to bin and rename it something like import_fisher.py?

File: util/importers/LDC97S62.py [hunk 25]
```diff
@@ -1,80 +1,43 @@
+#!/usr/bin/env python
```
Comment (kdavis-mozilla) [inline util/importers/LDC97S62.py]: Maybe move this to bin and rename it something like import_LDC97S62.py?

**Comment by kdavis-mozilla** [comment]
@reuben I'm done my run through.

File: DeepSpeech.py [hunk 26]
```diff
@@ -10,37 +9,41 @@
...
-from six.moves import zip, range, filter, urllib, BaseHTTPServer
-from threading import Thread, Lock
+
+# Hack: tf.app.flags only supports nargs='?', so we have to add our own method
+def DEFINE_list(flag_name, default_value, docstring):
+    tf.app.flags._global_parser.add_argument('--' + flag_name,
+                                             nargs='+',
+                                             help=docstring,
+                                             default=default_value,
+                                             type=str)
```
Comment (reuben) [inline DeepSpeech.py]: `os.pathsep` varies by OS so our bin/run-* scripts would require extra complexity to use the correct syntax, so I'm gonna go with a comma separated list for now.

File: util/data_set_helpers.py (hunk 15 above)
Comment (reuben) [inline util/data_set_helpers.py]: Yes, but we used to do:
```python
 return int(ceil(float(len(self._txt_files)) / float(self.batch_size)))
```
So:
```python
>>> from math import ceil
>>> int(ceil(float(10) / float(6)))
2
>>> from __future__ import division
>>> int(ceil(10 / 6))
2
```

File: DeepSpeech.py (hunk 16 above)
Comment (reuben) [inline DeepSpeech.py]: Python 3's `range` is Python 2's `xrange`. We could use `range` everywhere and pay a (probably negligible) performance price on Python 2, but why bother when we can just import `six.moves.range` instead? We already have the six dependency anyway.

File: util/data_set_helpers.py (hunk 15 above)
Comment (kdavis-mozilla) [inline util/data_set_helpers.py]: How is what we used to do relevant?

Comment (reuben) [inline util/data_set_helpers.py]: We used to force true division with the float() conversions, now we force it with `from __future__ import division`. The end result is the same, the only difference is that the float() calls went away.

Comment (kdavis-mozilla) [inline util/data_set_helpers.py]: Oh I see you added
```python
from __future__ import division
```
and changed from 
```python
int(ceil(float(len(self._txt_files)) / float(self.batch_size)))
```
to 
```python
int(ceil(len(self._filelist) / self.batch_size))
```
I thought you just added the import and the code was originally
```python
int(ceil(len(self._filelist) / self.batch_size))
```

**Comment by kdavis-mozilla** [comment]
@reuben FYI @andrenatal is on PTO until Wednesday. So you'll have to wait until then for his review.

**Comment by reuben** [comment]
@andrenatal I'm gonna land without your review for now, but feel free to add comments here anytime and I'll address them in follow-ups. We're not currently using Switchboard so it's not urgent.

**Pull request merged**